File: posts/2018/算法/src/mergeSort.js
class MergeSort {
  constructor (arr) {
    this.arr = arr
  }
  

  merge (left, mid, right) {
    let i = left
    let j = mid + 1
    const aux = []
    for (let k = left; k <= right; k++) {
      aux[k] = this.arr[k]
    }

    for (let k = left; k <= right; k++) {
      if (i > mid) this.arr[k] = aux[j++]
      else if (j > right) this.arr[k] = aux[i++]
      else if (aux[i] >= aux[j]) this.arr[k] = aux[j++]
      else this.arr[k] = aux[i++]
    }
  }

  sort (left, right) {
    if (left >= right) return
    
    const mid = left + Number.parseInt((right - left) / 2)
    this.sort(left, mid)
    this.sort(mid + 1, right)
    this.merge(left, mid, right)
  }
}


const array = [5,2,4,7,1,3,2,6]
const left = 0
const right = array.length - 1
const mergeSort = new MergeSort(array)
mergeSort.sort(0, right)
console.log(mergeSort.arr)

File: posts/2018/算法/src/go/mergeSort.go
package main

import (
	"fmt"
)

func mergeSort(arr []int, left, right int) {
	if left >= right {
		return
	}

	mid := left + (right-left)/2
	mergeSort(arr, left, mid)
	mergeSort(arr, mid+1, right)
	merge(arr, left, mid, right)
}

func merge(arr []int, left, mid, right int) {
	var tmp []int
	i := left
	j := mid + 1

	for i <= mid && j <= right {
		if arr[i] <= arr[j] {
			tmp = append(tmp, arr[i])
			i++
		} else {
			tmp = append(tmp, arr[j])
			j++
		}
	}

	start := i
	end := mid
	if j <= right {
		start = j
		end = right
	}
	for start <= end {
		tmp = append(tmp, arr[start])
		start++
	}

	for i := 0; i <= right-left; i++ {
		arr[left+i] = tmp[i]
	}

	fmt.Println(arr)
}

func main() {
	arr := []int{5, 2, 4, 7, 1, 3, 2, 6}
	mergeSort(arr, 0, len(arr)-1)
}
